src_c/_base_audio.c

```
@@ -0,0 +1,1137 @@
+// AUDIO.AUDIODEVICE CLASS
+// ***************************************************************************
+
+// The documentation says heap types need to support GC, so we're implementing
+// traverse even though the object has no explicit references.
+static int
+adevice_state_traverse(PyObject *op, visitproc visit, void *arg)
+{
+    // Visit the type
+    Py_VISIT(Py_TYPE(op));
+    return 0;
+}
+
+static void
+adevice_state_dealloc(PGAudioDeviceStateObject *self)
+{
+    // Only close devices that have been opened.
+    // (logical devices, not physical)
+    if (!SDL_IsAudioDevicePhysical(self->devid)) {
+        SDL_CloseAudioDevice(self->devid);
+    }
+    PyObject_GC_UnTrack(self);
+    PyTypeObject *tp = Py_TYPE(self);
+    freefunc free = PyType_GetSlot(tp, Py_tp_free);
+    free(self);
+    Py_DECREF(tp);
+}
+
+static PyMemberDef adevice_state_members[] = {
+    {"id", Py_T_UINT, offsetof(PGAudioDeviceStateObject, devid), Py_READONLY,
+     NULL},
+    {NULL} /* Sentinel */
+};
+
+static PyType_Slot adevice_state_slots[] = {
+    {Py_tp_members, adevice_state_members},
+    {Py_tp_traverse, adevice_state_traverse},
+    {Py_tp_dealloc, adevice_state_dealloc},
+    {0, NULL}};
+
+static PyType_Spec adevice_state_spec = {
+    .name = "AudioDeviceState",
+    .basicsize = sizeof(PGAudioDeviceStateObject),
+    .itemsize = 0,
+    .flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+    .slots = adevice_state_slots};
+
+static PyObject *
+pg_audio_is_audio_device_playback(PyObject *module, PyObject *arg)
+{
+    // SDL_IsAudioDevicePlayback
+    // arg: PGAudioDeviceStateObject
+
+    SDL_AudioDeviceID devid = ((PGAudioDeviceStateObject *)arg)->devid;
+    if (SDL_IsAudioDevicePlayback(devid)) {
+        Py_RETURN_TRUE;
+    }
+    Py_RETURN_FALSE;
+}
+
+static PyObject *
+pg_audio_get_audio_device_name(PyObject *module, PyObject *const *args,
+                               Py_ssize_t nargs)
```

if this always expects exactly one arg why not just use `METH_O` convention? As of now this just silently ignores extra args, and must have ub for no arg?

---

Yep that should be METH_O. (It’s one of the first functions I wrote and I didn’t adjust it to match the later style). I’m away from my computer for a few days so I will make the change when I am back or you/anyone could push that.

I intentionally omitted all argument error handling, although I used safe/stable ABI operations for all Pyobjects once received. The idea is that if checks can be moved into the Python VM that JIT Python might be smart enough to eliminate them and lead to higher performance. Not that performance is particularly important for this module. 

Maybe a simple and cheap check to do would be to check the number of args is as expected?